sdks/python/apache_beam/runners/portability/fn_api_runner_test.py

```diff
@@ -1083,6 +1131,34 @@ def __reduce__(self):
 def _unpickle_element_counter(name):
   return _pickled_element_counters[name]
 
+class EventRecorder(object):
+  """Used to be registered as a callback in bundle finalization.
+
+  The reason why records are written into a tmp file is, the in-memory dataset
+  cannot keep callback records when passing into one DoFn.
+  """
+  def __init__(self, tmp_dir):
+    self.tmp_dir = tmp_dir
+    self.file_path = os.path.join(self.tmp_dir, uuid.uuid4().hex + '.txt')
```

Comment:
This is not safe to concurrent access. Instead, create a new (random) path like you're doing here in record(), and have events() list all files and return the contents of each, and shutil.rmtree to clean them all up.

Reply:
If the current access means multi threads writes to the same file, I would prefer using lock, since If I create a new file for every record call, why there is a need to write content to that file?

Reply:
If we want to run these tests on a distributed runner (including Flink/Spark/... on a local machine), there'll be no distributed lock to grab. 

You would write content to that file because the content comes from an external source, and you have not control over it other than that it's a bunch of bytes.

Reply:
Fixed in https://github.com/apache/beam/pull/7937/commits/c5222da50865962311b5f475c9260864452acd0c